File: src/org/yiouli/leetcode/medium/AddTwoNumbers.java
/**
 * @author Yiou Li
 */
package org.yiouli.leetcode.medium;

/**
 * You are given two linked lists representing two non-negative numbers. The digits 
 * are stored in reverse order and each of their nodes contain a single digit. Add 
 * the two numbers and return it as a linked list.
 * Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
 * Output: 7 -> 0 -> 8
 * 
 * @see <a href="https://oj.leetcode.com/problems/add-two-numbers/">Add Two Numbers on LeetCode</a>
 */
public class AddTwoNumbers {
	public class ListNode {
		int val;
		ListNode next;
		ListNode(int x) {
			val = x;
			next = null;
		}
	}
	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		if (l1 == null) return l2;
		if (l2 == null) return l1;
		int carry = 0;
		ListNode head = new ListNode(0), prev = head;
		while (l1 != null || l2 != null) {
			int sum = carry;
			if (l1 != null) {
				sum += l1.val;
				l1 = l1.next;
			}
			if (l2 != null) {
				sum += l2.val;
				l2 = l2.next;
			}
			carry = sum / 10;
			prev.next = new ListNode(sum % 10);
			prev = prev.next;
		}
		if (carry != 0) prev.next = new ListNode(carry);
		return head.next;
	}
}

File: src/py/AddTwoNumbers.py
"""
Description

You are given two non-empty linked lists representing two non-negative integers.
The digits are stored in reverse order and each of their nodes contain a single digit.
Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        sentinel = current = ListNode()
        carry = 0
        while l1 is not None and l2 is not None:
            total = carry + l1.val + l2.val
            current.next = ListNode(total % 10)
            current = current.next
            carry = total // 10
            l1 = l1.next
            l2 = l2.next
        remain = l1 or l2
        while remain is not None:
            total = carry + remain.val
            current.next = ListNode(total % 10)
            current = current.next
            carry = total // 10
            remain = remain.next
        if carry > 0:
            current.next = ListNode(carry)
        return sentinel.next
            